src/PresentationalComponents/TagsToolbar/TagsToolbar.js

```diff
@@ -104,24 +141,37 @@ const TagsToolbar = ({ selectedTags, setSelectedTags }) => {
         >
             <InputGroup className='tags-filter-group'>
                 <TextInput
+                    aria-label='tags-filter-input'
                     placeholder={intl.formatMessage(messages.filterTagsInToolbar)}
                     value={searchText}
                     onChange={setSearchText}
                 />
                 <SearchIcon className='tags-filter-search-icon'/>
             </InputGroup>
             <Divider key="inline-filter-divider" />
-            {tags.slice(0, showMoreCount || tags.length).map(item =>
-                <SelectOption key={item} value={`${item}`}>
-                    <Tooltip content={`${decodeURIComponent(item)}`} position={TooltipPosition.right}>
-                        <span>{`${decodeURIComponent(item)}`}</span>
-                    </Tooltip>
-                </SelectOption>
+            {tags.map((group, index) =>
+                <SelectGroup key={`group${index}`} label={group.source}>
+                    {group.data.map((tag, tagIndex) =>
+                        <span key={tagIndex} className='tags-select-group'>
+                            <SelectOption value={tag.id} isChecked={
+                                selectedTags.find(selection => selection === tag.id)
+                            }>
+                                <Tooltip content={`${tag.id}`} position={TooltipPosition.right}>
+                                    <span>{`${decodeURIComponent(tag.value)}`}</span>
+                                </Tooltip>
+                            </SelectOption>
+                            <Badge className='tags-select-badge'> {tag.count} </Badge>
+                        </span>
+                    )}
+                </SelectGroup>
             )}
             <Button key='manage all tags'
-                variant='link' onClick={(toggleModal) => setManageTagsModalOpen(toggleModal)}>
-                {(showMoreCount > 0 && tags.length > showMoreCount) ?
-                    intl.formatMessage(messages.countMoreTags, { count: tags.length - showMoreCount }) : intl.formatMessage(messages.manageTags)
+                variant='link' onClick={() => setManageTagsModalOpen(true)}>
+                {(showMoreCount > 0 && tags.flatMap(source => source.data).length >= showMoreCount) ?
```

Comment:
`tags.flatMap(source => source.data).length` happens twice, any way we can make it happen once?